### ISSUE TITLE
Missing support for vault.hcl file

### ISSUE THREAD

**Issue by deebhatia** [jsok/puppet-vault#183]
The config file path has been changed from `config.json` to `vault.hcl`. Hence, we should update the file prefix as vault.hcl.

**Comment by mattpascoe** [comment]
I too am running into the vault.hcl issue.  I'm using the .deb packages that are provided by hashicorp repos (https://apt.releases.hashicorp.com/).  They provide a systemd startup that expects vault.hcl to exist.

The challenge here of course is that I would have to now provide my own vault.service file for systemd to get it to use the .json file.

It looks like hashicorp is moving to .hcl as their default so it would seem that this module might want to move that direction.  If we could provide an option for config format and have the templates to build either a .hcl or .json file that would be the most flexible way.  As it stands now, I feel that the hashicorp provided .deb files and this puppet module are fighting each other which requires extra work to get them to play nice.

Here is the current vault.service file that the hashicorp .deb (version 1.6.0) file lays down:
`[Unit]
Description="HashiCorp Vault - A tool for managing secrets"
Documentation=https://www.vaultproject.io/docs/
Requires=network-online.target
After=network-online.target
ConditionFileNotEmpty=/etc/vault.d/vault.hcl
StartLimitIntervalSec=60
StartLimitBurst=3

[Service]
User=vault
Group=vault
ProtectSystem=full
ProtectHome=read-only
PrivateTmp=yes
PrivateDevices=yes
SecureBits=keep-caps
AmbientCapabilities=CAP_IPC_LOCK
CapabilityBoundingSet=CAP_SYSLOG CAP_IPC_LOCK
NoNewPrivileges=yes
ExecStart=/usr/bin/vault server -config=/etc/vault.d/vault.hcl
ExecReload=/bin/kill --signal HUP $MAINPID
KillMode=process
KillSignal=SIGINT
Restart=on-failure
RestartSec=5
TimeoutStopSec=30
StartLimitInterval=60
StartLimitBurst=3
LimitNOFILE=65536
LimitMEMLOCK=infinity

[Install]
WantedBy=multi-user.target`

**Comment by jsok** [comment]
Thanks for pointing out @mattpascoe. That's a very different situation to just renaming the config file.

> As it stands now, I feel that the hashicorp provided .deb files and this puppet module are fighting each other which requires extra work to get them to play nice.

Naturally, going forward it sounds like we're going to need to keep them in sync, e.g. if the systemd unit file decides to change the config file location one day without warning.

To be clear, there's no reason this module needs to generate the config in HCL format. Vault will be smart enough to decode the config file irrespective of the file extension, and it's much simpler to machine generate JSON (`stdlib` provides this) than HCL (which would require a bunch of custom code for no real advantage).

So the real issue here is how we support the path the Hashicorp deb packages expect the config to live at.
The options I can think of:

- Generate a systemd drop in which clears/updates the `ConditionFileNotEmpty` option and overwrites the `ExecStart`
- Overwrite `config_dir` and introduce a new `config_file` parameter so we can set the path to `/etc/vault.d/vault.hcl`

Open to hear other suggestions before we decide on the best approach.

**Comment by mattpascoe** [comment]
I agree with your assessment of using JSON being much simpler.  I'm perfectly fine with that as it does not look like hashicorp has any intentions of dropping support for that.

As far as the fix, I too am torn. 

I'd love to leave the hashicorp default path of `/etc/vault.d/vault.hcl`, but it is a bit confusing to have JSON contents in the HCL file.  Tho, if I'm using puppet to manage the file its already something I shouldnt be touching and is simple enough to read.  

I'm not sure the complexities of updating the current systemd file with proper paths, I assume it would be some sort of augeus type situation.  That would also work.

Another thought tho might be to leave the systemd file alone and simply just put a comment line only in `/etc/vault.d/vault.hcl`.  That way it matches the CondidtionFile.  I'm hoping the `/etc/vault.d` situation works such that you could have `/etc/vault.d/vault.json` and it would use that.  The question would be the `-config=` line in systemd.  would that actually pick up the other files in `/etc/vault.d`.. .....................

ok, instead of typing my idea I just tried it.. basically it does not work since their systemd file specifies the full path to the file itself.  They should have just used /etc/vault.d.  The document says:

>  -config=<string>
      Path to a configuration file or directory of configuration files. This
      flag can be specified multiple times to load multiple configurations.
      If the path is a directory, all files which end in .hcl or .json are
      loaded.

So, it still might be useful to implement it that way but would have been nice if hashicorp actually used the `.d` paradigm correctly.

Anyway, hope those thoughts help in some way.

**Comment by jsok** [comment]
> So, it still might be useful to implement it that way but would have been nice if hashicorp actually used the .d paradigm correctly.

That is a good point. I've been searching to see if the Hashicorp releases packaging code is on GitHub somewhere but coming up blank. I find it a bit odd that we can't see that code or raise PRs to potentially improve the systemd unit file.

To give you a rough idea of what a systemd drop-in would look like (after being templated):

`/etc/systemd/system/vault.service.d/puppet-vault.conf`:

```ini
[Unit]
# Clear the check for vault.hcl
ConditionFileNotEmpty=
ConditionFileNotEmpty=/etc/vault.d/config.json

[Service]
ExecStart=/usr/bin/vault server -config=/etc/vault.d/config.json
```

**Comment by mattpascoe** [comment]
Yea I agree, I'm not seeing a good way to ask them to update the package configuration like that.  

ok yea I got what you are going with on the systemd dropin.  I like that idea the best then.  I think given the current options it should be the cleanest most portable way to accomplish a consistent method to manage it.  I'm not certain how that might effect other methods your module supports (various package types and or non package methods) but I think this would certainly fix my issues.

Heck this would even give the possibility to allow extra options to be added to the ExecStart line if the user wanted to add them.  Basically I'm thinking you create a `systemd_execflags` option that would be templated in to the end of the ExecStart line?  defaults to empty of course.

**Comment by mattpascoe** [comment]
I manually adapted some of this on my test box.  I think it will work just fine.  Basically I updated `config_dir` to `/etc/vault.d` and then manually created a `/etc/systemd/system/vault.service.d/puppet-vault.conf` file.  I had to add an `ExecStart=` line to clear it out as well for it to work.

So I think the idea is certainly viable.

**Comment by linuxmail** [comment]
Hello,

there is one issue, with the HCL vs. JSON stuff ...  its not in Sync on Vault code. I have the problem, that I couldn't add service registration (as vault extra_config), because the created json was not correctly parsed from the vault. Ok, it was [fixed](https://github.com/hashicorp/vault/issues/17012), but I assume, it will break in the future too, because they use hcl as preferred config file.
But anyway ..  I can't get it working ...

```2022-11-04T11:35:23.879+0100 [WARN]  unknown or unsupported field consul found in configuration at /etc/vault/config.json
2022-11-04T11:35:23.879+0100 [INFO]  proxy environment: http_proxy="" https_proxy="" no_proxy=""
2022-11-04T11:35:23.887+0100 [INFO]  core: Initializing version history cache for core
2022-11-04T11:35:23.889+0100 [INFO]  core: raft retry join initiated
2022-11-04T11:35:23.890+0100 [ERROR] service_registration.consul: error running service registration: failed to find a host:port in redirect address "": missing port in address
```
cu denny
